src/libraries/UniswapV3Lib.sol

```diff
@@ -0,0 +1,259 @@
+        IRateLimits rateLimits;
+        bytes32     rateLimitId;
+        address     pool;
+        uint256     deadline;
+    }
+
+    struct SwapParams {
+        UniswapV3PoolParams poolParams;
+        address             router;
+        address             tokenIn;
+        uint256             amountIn;
+        uint256             minAmountOut;
+        uint24              tickDelta; // The maximum that the tick can move by after completing the swap; type(uint24).max for no limit
+        uint256             maxSlippage;
+    }
+
+    struct SwapCache {
+        IUniswapV3PoolLike pool;
+        address            token0;
+        address            token1;
+        address            tokenOut;
+        uint160            sqrtPriceLimitX96;
+        uint256            twapExpectedOut;
+    }
+
+    /**********************************************************************************************/
+    /*** External functions                                                                     ***/
+    /**********************************************************************************************/
+
+    // Rate limit decreased by value of token1 
+    function swap(UniV3Context calldata context, SwapParams calldata params) external returns (uint256 amountOut) {
+        SwapCache memory cache = _populateSwapCache(context, params);
+
+        require(
+            params.tokenIn == cache.token0 || params.tokenIn == cache.token1,
+            "UniswapV3Lib/invalid-token-pair"
+        );
+        require(params.maxSlippage > 0, "UniswapV3Lib/max-slippage-not-set");
+        require(params.tickDelta <= params.poolParams.swapMaxTickDelta, "UniswapV3Lib/invalid-max-tick-delta");
+
+        uint256 minOutBySlippage = cache.twapExpectedOut * params.maxSlippage / 1e18;
+
+        require(params.minAmountOut >= minOutBySlippage, "UniswapV3Lib/min-amount-not-met");
+
+        _approve(context.proxy, params.tokenIn, params.router, params.amountIn);
+
+        amountOut = _callSwap(context, params, cache);
+
+        context.rateLimits.triggerRateLimitDecrease(
+            RateLimitHelpers.makeAssetDestinationKey(context.rateLimitId, params.tokenIn, context.pool),
+            _scaleTo1e18(params.amountIn, IERC20Metadata(params.tokenIn).decimals())
+        );
+    }
+
+    /**********************************************************************************************/
+    /*** Helper functions                                                                       ***/
+    /**********************************************************************************************/
+    
+    //-- Swap helper functions
+    function _populateSwapCache(UniV3Context calldata context, SwapParams calldata params) internal view returns (SwapCache memory) {
+        IUniswapV3PoolLike pool         = IUniswapV3PoolLike(context.pool);
+        address token0                  = pool.token0();
+        address token1                  = pool.token1();
```

Comment:
Instead of declaring individual variables here, can you name the swap cache memory return, and then populate token0, token1, tokenOut, etc., in the pool as you go?

Reply:
https://github.com/grove-labs/grove-alm-controller/pull/54/commits/ed834be825e34196051409eb8ef9c59b729ece19